gateway/src/apicast/policy/caching/policy.lua

```diff
@@ -38,13 +44,39 @@ local function resilient_handler(cache, cached_key, response, ttl)
   end
 end
 
+local function handle_500_allow_mode(cache, cached_key, ttl)
+  local current_value = cache:get(cached_key)
+  local cached_4xx = current_value and current_value >= 400 and current_value < 500
+
+  if not cached_4xx then
+    ngx.log(ngx.WARN, 'Backend seems to be unavailable. "Allow" mode is ',
+                      'enabled in the cache policy, so next request will be ',
+                      'authorized')
+    cache:set(cached_key, 200, ttl)
```

Comment:
What about the race condition between read and write?

Reply:
Good point. Unfortunately, I don't see any 'compare-and-set' operation in the API of ngx.shared.dict: https://github.com/openresty/lua-nginx-module#ngxshareddict 

I see that there's this library: https://github.com/openresty/lua-resty-lock but that would mean introducing a dependency just for this policy.

Reply:
@davidor lua-resty-lock is part of the openresty bundle, so it is part of stdlib. 

I think we could take some time and think about how to do this without the lock anyway. Possibly using the `add` operation. If the stdict value would be nil, we could `add` 200. 
Not sure how exactly that should work, but I see it as the only option without using the lock because it is the only compare-and-set atomic (compare to nil) operation on shdict.

Reply:
I think that `add` would work in this case. There are only 3 possibilities:
- If there was a 2XX, we don't need to overwrite it with the same thing.
- If there was a 4XX, we don't need to write anything.
- If it was nil, we need to write a 200.

Let me try this.

Reply:
Solved in the new commit @mikz

Reply:
👍